### PR TITLE
plugin handler: properly handle snapcraftctl errors

### PR DESCRIPTION
**Pull request by cjp256**
Address issues including:

- snapcraftctl operations always exiting with 0, even for failure
  cases.

- snapcraft override scripts continue to run even after
  parent snapcraft process fails.  This is because `set -e` does
  not catch the error when snapcraftctl returns 0, and nothing
  is otherwise done to terminate the script.

Given the nature of these symptoms, this is likely the cause of
a number of other seemingly unrelated issues.

These issues occur because _handle_builtin_function() only
catches exceptions of type errors.ScriptletBaseError,
ignoring other cases that may result from the build/pull operations,
such as NodejsPluginMissingPackageJsonError, or
SnapcraftPluginBuildError.

Instead of trying to manage exceptions in _handle_builtin_function(),
catch _any_ unhandled exception in _run_scriplet()'s try block. This
simplifies the more complex return value and feedback handling.

Don't print an error from the feedback channel in snapcraftctl,
the parent process will take care of it.  This is necessary
because the error that is now seen by snapcraftctl would lead
to redundant output.  Just exit with -1 when an error is
read from the feedback fifo.

LP: #1898922

Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

**Comment by sergiusens** [comment]
Can you add a spread test to the general suite here?

File: tests/spread/general/scriptlets/task.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  rm -f ./*.snap
+
+  restore_yaml "snap/snapcraft.yaml"
+
+execute: |
+  #shellcheck source=tests/spread/tools/snapcraft-yaml.sh
+  . "$TOOLS_DIR/snapcraft-yaml.sh"
+  cd "$SNAP_DIR"
+
+
```
Comment (sergiusens) [inline tests/spread/general/scriptlets/task.yaml:26]: random question about the amount of empty lines and the thought process put into it

Comment (cjp256) [reply tests/spread/general/scriptlets/task.yaml:26]: There does seem to be an extra line there for no reason. Just an accident.
```suggestion
```

File: tests/spread/general/scriptlets/task.yaml [hunk 2]
```diff
@@ -9,20 +9,16 @@ prepare: |
...
 
-  restore_yaml "snap/snapcraft.yaml"
-
-execute: |
   #shellcheck source=tests/spread/tools/snapcraft-yaml.sh
   . "$TOOLS_DIR/snapcraft-yaml.sh"
-  cd "$SNAP_DIR"
+  restore_yaml "snap/snapcraft.yaml"
 
+execute: |x
```
Comment (sergiusens) [inline tests/spread/general/scriptlets/task.yaml:20]: there's an x here now :-)

Comment (sergiusens) [inline tests/spread/general/scriptlets/task.yaml:20]:
```suggestion
execute: |
```

**Review by sergiusens — APPROVED** [review]

**Pull request merged**